File: codequest-api/app/schemas/agent.py
"""智能体工坊相关的 Pydantic 模型"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class EnergyRating(BaseModel):
    """能量评级 - 四维度评分"""
    understanding: int = Field(0, ge=0, le=5, description="理解力")
    implementation: int = Field(0, ge=0, le=5, description="实现力")
    optimization: int = Field(0, ge=0, le=5, description="优化力")
    creativity: int = Field(0, ge=0, le=5, description="创造力")


class NeuronNodeBrief(BaseModel):
    """地图上的节点简要信息"""
    id: int
    title: str
    slug: str
    track: str
    section: Optional[str] = None
    difficulty: Optional[str] = None
    order: int = 0
    xp_reward: int = 10
    prerequisites: list[int] = []
    status: Optional[str] = None
    energy_score: Optional[int] = None

    model_config = {"from_attributes": True}


class NeuronNodeResponse(BaseModel):
    """节点详情"""
    id: int
    title: str
    slug: str
    description: Optional[str] = None
    content: Optional[str] = None
    order: int = 0
    difficulty: Optional[str] = None
    xp_reward: int = 10
    track: str
    section: Optional[str] = None
    starter_code: Optional[str] = None
    hint: Optional[str] = None
    prerequisites: list[int] = []
    energy_levels: Optional[EnergyRating] = None
    status: Optional[str] = None
    energy_score: Optional[int] = None
    energy_detail: Optional[EnergyRating] = None
    best_code: Optional[str] = None
    attempts: int = 0

    model_config = {"from_attributes": True}


class NeuralMapEdge(BaseModel):
    """地图连线"""
    source: int
    target: int


class NeuralMapResponse(BaseModel):
    """神经元网络地图"""
    nodes: list[NeuronNodeBrief]
    edges: list[NeuralMapEdge]


class AgentProgressResponse(BaseModel):
    """用户智能体工坊进度"""
    id: int
    node_id: int
    status: str
    energy_score: Optional[int] = None
    energy_detail: Optional[EnergyRating] = None
    best_code: Optional[str] = None
    attempts: int = 0
    completed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class AgentSubmitRequest(BaseModel):
    """提交请求"""
    code: str = Field(..., min_length=1, max_length=50000, description="提交的代码")


class AgentSubmitResult(BaseModel):
    """提交结果"""
    submission_id: str
    status: str
    score: int = 0
    energy: EnergyRating = Field(default_factory=EnergyRating)
    energy_score: int = 0
    stdout: str = ""
    stderr: str = ""
    execution_time: float = 0
    encouragement_message: str = ""
    xp_earned: int = 0
    test_results: list[dict] = []
    ai_analysis: str = ""


class TrackOverview(BaseModel):
    """主线概览"""
    track: str
    name: str
    description: str
    total_nodes: int = 0
    completed_nodes: int = 0
    sections: list[str] = []

File: HelloWorld-web/src/api/agent.ts
import apiClient from './client'

export interface NeuronNode {
  id: number
  title: string
  slug: string
  description: string | null
  track: string
  section: string
  difficulty: string
  xp_reward: number
  prerequisites: number[]
  energy_score: number
  status: string
}

export interface NeuralEdge {
  source: number
  target: number
}

export interface NeuralMapData {
  nodes: NeuronNode[]
  edges: NeuralEdge[]
  tracks: TrackOverview[]
}

export interface TrackOverview {
  track: string
  name: string
  description: string
  total_nodes: number
  completed_nodes: number
}

export interface AgentNodeDetail {
  id: number
  title: string
  slug: string
  description: string | null
  content: string | null
  track: string
  section: string
  difficulty: string
  xp_reward: number
  starter_code: string | null
  hint: string | null
  prerequisites: number[]
  energy_score: number
  energy_detail: { understanding: number; implementation: number; optimization: number; creativity: number }
  status: string
  attempts: number
}

export interface AgentSubmitResult {
  status: string
  score: number
  energy_score: number
  energy_detail: { understanding: number; implementation: number; optimization: number; creativity: number }
  stdout: string | null
  stderr: string | null
  ai_analysis: string | null
  xp_earned: number
}

export async function getNeuralMap(): Promise<NeuralMapData> {
  const { data } = await apiClient.get<NeuralMapData>('/agent/map')
  return data
}

export async function getAgentNode(nodeId: number): Promise<AgentNodeDetail> {
  const { data } = await apiClient.get<AgentNodeDetail>(`/agent/nodes/${nodeId}`)
  return data
}

export async function submitAgentCode(nodeId: number, code: string): Promise<AgentSubmitResult> {
  const { data } = await apiClient.post<AgentSubmitResult>(`/agent/nodes/${nodeId}/submit`, { code })
  return data
}

export async function getAgentTracks(): Promise<TrackOverview[]> {
  const { data } = await apiClient.get<TrackOverview[]>('/agent/tracks')
  return data
}
